test/Microsoft.ApplicationInsights.AspNet.Tests/TelemetryInitializers/OperationNameTelemetryInitializerTests.cs

```diff
@@ -72,16 +71,107 @@ public void InitializeSetsTelemetryOperationNameToMethodAndPath()
-            ac.HttpContext.RequestServices = new TestServiceProvider(new List<object>() { new RequestTelemetry() });
-            var initializer = new OperationNameTelemetryInitializer(ac);
+            var contextAccessor = HttpContextAccessorHelper.CreateHttpContextAccessor(telemetry, null);
+
+            var initializer = new OperationNameTelemetryInitializer(contextAccessor);
 
             initializer.Initialize(telemetry);
 
             Assert.Equal("GET /Test", telemetry.Name);
         }
+
+        [Fact]
+        public void InitializeSetsTelemetryOperationNameToControllerFromActionContext()
+        {
+            var actionContext = new ActionContext();
+            actionContext.RouteData = new Microsoft.AspNet.Routing.RouteData();
+            actionContext.RouteData.Values.Add("controller", "home");
+
+            var contextAccessor = HttpContextAccessorHelper.CreateHttpContextAccessor(new RequestTelemetry(), actionContext);
+
+            var initializer = new OperationNameTelemetryInitializer(contextAccessor);
+
+            var telemetry = new EventTelemetry();
+            initializer.Initialize(telemetry);
+
+            Assert.Equal("GET home", telemetry.Context.Operation.Name);
+        }
+
+        [Fact]
+        public void InitializeSetsTelemetryOperationNameToControllerAndActionFromActionContext()
+        {
+            var actionContext = new ActionContext();
+            actionContext.RouteData = new Microsoft.AspNet.Routing.RouteData();
+            actionContext.RouteData.Values.Add("controller", "account");
+            actionContext.RouteData.Values.Add("action", "login");
+
+            var contextAccessor = HttpContextAccessorHelper.CreateHttpContextAccessor(new RequestTelemetry(), actionContext);
+
+            var initializer = new OperationNameTelemetryInitializer(contextAccessor);
+
+            var telemetry = new EventTelemetry();
+            initializer.Initialize(telemetry);
+
+            Assert.Equal("GET account/login", telemetry.Context.Operation.Name);
+        }
+
+        [Fact]
+        public void InitializeSetsTelemetryOperationNameToControllerAndActionAndParameterFromActionContext()
+        {
+            var actionContext = new ActionContext();
+            actionContext.RouteData = new Microsoft.AspNet.Routing.RouteData();
+            actionContext.RouteData.Values.Add("controller", "account");
+            actionContext.RouteData.Values.Add("action", "login");
+            actionContext.RouteData.Values.Add("parameter", "myName");
+
+            var contextAccessor = HttpContextAccessorHelper.CreateHttpContextAccessor(new RequestTelemetry(), actionContext);
+
+            var initializer = new OperationNameTelemetryInitializer(contextAccessor);
+
+            var telemetry = new EventTelemetry();
+            initializer.Initialize(telemetry);
+
+            Assert.Equal("GET account/login [parameter]", telemetry.Context.Operation.Name);
```

Comment:
I don't quite see how replacing `Assert.Equal` with `Assert.Contains` or `Assert.EndsWith` would overcomplicate tests. :confused: Here is another question - if the asserts and tests are fine the way they are, why do we want to have 4 separate tests with mostly [duplicate code](http://xunitpatterns.com/Test%20Code%20Duplication.html) instead of just one - `InitializeSortsParameters`?